### PR TITLE
Naive, Traditional ML, Deep Learning Approaches for Anime Character Generation

### PR DESCRIPTION
**Pull request by billyae**
## Summary
This PR delivers a complete Anime Character Generation project with three distinct style‑transfer approaches, their corresponding evaluation scripts, and two Streamlit demo apps. It lays the groundwork for training, validating, and interactive inference of:

- Naive K‑Means Quantization + Edge Overlay 
- Traditional ML (RandomForest over RGB patches)
- Diffusion + ControlNet (fine‑tuned Stable Diffusion pipeline)

## What’s Added

1. Naive Color‑Quantization Pipeline

    `naive_train.py`: Learns a global 32‑color palette via K‑means on anime labels, applies palette quantization + Canny edges to human photos.     
    `naive_evaluation.py`: Restyles test images, computes average CLIP similarity to true anime targets.   

2. Traditional ML Pipeline
    `traditional_ml_train.py`: Builds patch‐based feature dataset, trains a RandomForestRegressor, reports MSE & R², and saves model.       
    `traditional_ml_evaluation.py`: Loads the trained RF model, reconstructs test images, computes average CLIP cosine similarity against ground truth.     

3. Diffusion + ControlNet Pipeline  

    `diffusion_train.py`: Fine‑tunes a Stable Diffusion pipeline with ControlNet conditioning on human–anime pairs; freezes all but ControlNet; uses mixed‑precision via accelerate.    
    `diffusion_evaluation.py`: Loads the checkpoint, generates test outputs, and measures average CLIP similarity.    

4. Streamlit Demo Apps

    `app_naive.py`: Interactive web UI for the naive palette‑and‑edges method.     
    `app_dl.py`: Web UI for the fine‑tuned diffusion model; button to download the pretrained checkpoint from Google Drive and real‑time anime synthesis.   

## Evaluation Strategy

All three methods leverage CLIP‑based cosine similarity to quantify perceptual style matching between generated and ground‑truth anime images.

## How to Run

Install dependencies: `pip install requirements.txt`  
Train each approach:
`python traditional_ml_train.py`
`python naive_train.py`
`python diffusion_train.py`

Evaluate:
`python traditional_ml_evaluation.py`
`python naive_evaluation.py`
`python diffusion_evaluation.py`

Launch Demos:

`streamlit run app_naive.py`
`streamlit run app_dl.py`

## Next Steps

- Add CI checks for dataset structure & basic smoke tests.
- Integrate these scripts into a unified CLI or notebook.

### CODE REVIEW

**Comment by billyae** [comment]
Well structured code with good comments! Modular code with detailed docstrings in each function.
